### PR TITLE
Decision engine: Call driver pre-upgrade hooks

### PR DESCRIPTION
**Pull request by sc68cal**
These hooks allow an upgrade driver to do any prep work before a service
or host is upgraded.

### CODE REVIEW

File: kostyor/upgrades/engine.py [hunk 1]
```diff
@@ -152,17 +152,21 @@ def start(self):
         # keystone, then with nova, and so on. See iteration details in
         # get_controllers() docstring.
         for host in get_controllers(hosts):
+            self.driver.pre_host_upgrade_hook(host)
```
Comment (ikalnitsky) [inline kostyor/upgrades/engine.py]: Maybe submit returned value to tasks? So driver developer can return a certain Celery task/chain/group to do some long-running stuff?

Comment (sc68cal) [inline kostyor/upgrades/engine.py]: Can you elaborate some more on this?

File: kostyor/upgrades/engine.py [hunk 2]
```diff
@@ -152,17 +152,21 @@ def start(self):
...
             for service in iterservices(host):
+                self.driver.pre_service_upgrade_hook(service)
```
Comment (sc68cal) [inline kostyor/upgrades/engine.py]: The point of these hooks is that there _may_ be things that a driver does before a service or host is upgraded. We don't know, since it'd be up to the upgrade driver. This just gives the upgrade driver information about what is going on. 

I don't think putting everything inside of `start_upgrade` - including pre/post is a good practice.

File: kostyor/upgrades/engine.py (hunk 1 above)
Comment (ikalnitsky) [inline kostyor/upgrades/engine.py]: I mean having

```python
subtasks.append(self.driver.pre_host_upgrade_hook(host))
```

just in case `pre_host_upgrade_hook` wants to execute some long-running task. I think if driver doesn't need any pre-host upgrade hook, it will return some `NoopTask`.  Otherwise, we are limited to have quick tasks in order to do not block the process.

File: kostyor/upgrades/engine.py (hunk 2 above)
Comment (ikalnitsky) [inline kostyor/upgrades/engine.py]: There are cons/pros in both approaches. I think my concern is having a lot of methods in the interface leads to confusing and complexity in maintaining. I like when there's only one way to do a thing. Your proposal allows both approaches.

File: kostyor/upgrades/engine.py (hunk 1 above)
Comment (sc68cal) [inline kostyor/upgrades/engine.py]: OK I see what you're saying now. If the pre and post hooks to return `Tasks` that'd probably be very useful. I agree.

File: kostyor/upgrades/engine.py (hunk 2 above)
Comment (sc68cal) [inline kostyor/upgrades/engine.py]: Right. The pre and post hooks are not decorated with `abc.abstractmethod` - so a driver does not need to provide an implementation of them, like they do for the `*_upgrade` methods. So they can just rely on the pre and post hooks that they inherit from `UpgradeDriver` - which are noop

File: kostyor/upgrades/strategy.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import abc
```
Comment (ikalnitsky) [inline kostyor/upgrades/strategy.py]: I think the strategy.py must be a separate commit (and maybe even another PR), since it's not about "Decision engine: Call driver pre-upgrade hooks"

File: kostyor/upgrades/engine.py (hunk 2 above)
Comment (ikalnitsky) [inline kostyor/upgrades/engine.py]: Ok, let's proceed with that then.

File: kostyor/upgrades/strategy.py (hunk 3 above)
Comment (sc68cal) [inline kostyor/upgrades/strategy.py]: Yes but as part of a pre-upgrade task, we may need to live migrate instances off a host. It is related to this work. So, I'm including it.

Comment (ikalnitsky) [inline kostyor/upgrades/strategy.py]: i think the key word here is **may**.
